src/mame/drivers/hp3478a.cpp

```diff
@@ -0,0 +1,630 @@
+	PORT_DIPNAME( 0x1f, 0x17, "HP-IB Bus Address" ) PORT_DIPLOCATION("DIP:1,2,3,4,5")
+	PORT_DIPSETTING(    0x00, "0" )
+	PORT_DIPSETTING(    0x01, "1" )
+	PORT_DIPSETTING(    0x02, "2" )
+	PORT_DIPSETTING(    0x03, "3" )
+	PORT_DIPSETTING(    0x04, "4" )
+	PORT_DIPSETTING(    0x05, "5" )
+	PORT_DIPSETTING(    0x06, "6" )
+	PORT_DIPSETTING(    0x07, "7" )
+	PORT_DIPSETTING(    0x08, "8" )
+	PORT_DIPSETTING(    0x09, "9" )
+	PORT_DIPSETTING(    0x0a, "10" )
+	PORT_DIPSETTING(    0x0b, "11" )
+	PORT_DIPSETTING(    0x0c, "12" )
+	PORT_DIPSETTING(    0x0d, "13" )
+	PORT_DIPSETTING(    0x0e, "14" )
+	PORT_DIPSETTING(    0x0f, "15" )
+	PORT_DIPSETTING(    0x10, "16" )
+	PORT_DIPSETTING(    0x11, "17" )
+	PORT_DIPSETTING(    0x12, "18" )
+	PORT_DIPSETTING(    0x13, "19" )
+	PORT_DIPSETTING(    0x14, "20" )
+	PORT_DIPSETTING(    0x15, "21" )
+	PORT_DIPSETTING(    0x16, "22" )
+	PORT_DIPSETTING(    0x17, "23" )
+	PORT_DIPSETTING(    0x18, "24" )
+	PORT_DIPSETTING(    0x19, "25" )
+	PORT_DIPSETTING(    0x1a, "26" )
+	PORT_DIPSETTING(    0x1b, "27" )
+	PORT_DIPSETTING(    0x1c, "28" )
+	PORT_DIPSETTING(    0x1d, "29" )
+	PORT_DIPSETTING(    0x1e, "30" )
+	PORT_DIPSETTING(    0x1f, "31" )
+	PORT_DIPNAME( 0x20, 0x00, "PWR ON SRQ" ) PORT_DIPLOCATION("DIP:6")
+	PORT_DIPSETTING(    0x00, "Disabled" )
+	PORT_DIPSETTING(    0x20, "Enabled" )
+	//0x40 unused
+	PORT_DIPNAME( 0x80, 0x00, "50/60Hz AC" ) PORT_DIPLOCATION("DIP:8")
+	PORT_DIPSETTING(    0x00, "60Hz" )
+	PORT_DIPSETTING(    0x80, "50Hz" )
+
+INPUT_PORTS_END
+
+/******************************************************************************
+ Machine Drivers
+******************************************************************************/
+
+void hp3478a_state::hp3478a(machine_config &config)
+{
+	auto &mcu(I8039(config, "maincpu", CPU_CLOCK));
+	mcu.set_addrmap(AS_PROGRAM, &hp3478a_state::i8039_map);
+	mcu.set_addrmap(AS_IO, &hp3478a_state::i8039_io);
+	mcu.p1_in_cb().set(FUNC(hp3478a_state::p1read));
+	mcu.p1_out_cb().set(FUNC(hp3478a_state::p1write));
+	mcu.p2_out_cb().set(FUNC(hp3478a_state::p2write));
+
+	NVRAM(config, "nvram", nvram_device::DEFAULT_ALL_0);
+
+	ADDRESS_MAP_BANK(config, m_iobank, 0);
+	m_iobank->set_map(&hp3478a_state::io_bank);
+	m_iobank->set_data_width(8);
+	m_iobank->set_addr_width(18);
+	m_iobank->set_stride(0x100);
+
+	WATCHDOG_TIMER(config, m_watchdog).set_time(attotime::from_ticks(3*5*(1<<19),CPU_CLOCK));
+
+	// video
+	config.set_default_layout(layout_hp3478a);
+}
+
+/******************************************************************************
+ ROM Definitions
+******************************************************************************/
+ROM_START( hp3478a )
+	ROM_REGION( 0x2000, "maincpu", 0 )
+	ROM_LOAD("rom_dc118.bin", 0, 0x2000, CRC(10097ced) SHA1(bd665cf7e07e63f825b2353c8322ed8a4376b3bd))	//main CPU ROM, can match other datecodes too
+
+	ROM_REGION( 0x100, "nvram", 0 )	/* Calibration RAM, battery-backed */
+	ROM_LOAD_OPTIONAL( "calram.bin", 0, 0x100, CRC(0))
```

Comment:
Would it make sense to provide a ROM with default calibration values for the CALRAM? 

What would calibration mean for an ideal/perfect device like an emulator? Are there canonical calibration values in the case of a "perfect" measurement instrument?

Reply:
A sample calibration ? Perhaps, yes. The CAL stores an offset and gain (simple linear a + bx form) for all modes and ranges, so I guess a "perfect" device would just have 0 everywhere. I'd  have to test if a 0-filled dump would pass the CAL checksum check.